### PR TITLE
Miner tweaks

### PR DESCRIPTION
**Pull request by gavofyork**
- Use queue to determine whether we're mining;
- Kick stale hash rates;
- Removed some panickers from `Client`.

### CODE REVIEW

File: ethcore/src/client/client.rs [hunk 1]
```diff
@@ -178,15 +178,15 @@ impl Client {
 		db_config.compaction = config.db_compaction.compaction_profile();
 		db_config.wal = config.db_wal;
 
-		let db = Arc::new(Database::open(&db_config, &path.to_str().unwrap()).expect("Error opening database"));
+		let db = Arc::new(try!(Database::open(&db_config, &path.to_str().unwrap()).map_err(|s| ClientError::Database(s))));
```
Comment (tomusdrw) [inline ethcore/src/client/client.rs]: `map_err(ClientError::Database)` should work as well.

File: rpc/src/v1/impls/eth.rs [hunk 2]
```diff
@@ -316,7 +316,7 @@ impl<C, S: ?Sized, M, EM> Eth for EthClient<C, S, M, EM> where
 	fn is_mining(&self, params: Params) -> Result<Value, Error> {
 		try!(self.active());
 		match params {
-			Params::None => to_value(&self.external_miner.is_mining()),
+			Params::None => to_value(&(take_weak!(self.miner).is_sealing())),
```
Comment (tomusdrw) [inline rpc/src/v1/impls/eth.rs]: Wouldn't using `miner.sealing_work.enabled` work?

File: ethcore/src/client/client.rs (hunk 1 above)
Comment (debris) [inline ethcore/src/client/client.rs]: if `ClientError` implements `From<DatabaseError>`, `map_err` is redundant

File: ethcore/src/client/client.rs [hunk 3]
```diff
@@ -178,15 +178,15 @@ impl Client {
...
 		let chain = Arc::new(BlockChain::new(config.blockchain, &gb, db.clone()));
 		let tracedb = Arc::new(try!(TraceDB::new(config.tracing, db.clone(), chain.clone())));
 
 		let mut state_db = journaldb::new(db.clone(), config.pruning, DB_COL_STATE);
 		if state_db.is_empty() && spec.ensure_db_good(state_db.as_hashdb_mut()) {
 			let batch = DBTransaction::new(&db);
-			state_db.commit(&batch, 0, &spec.genesis_header().hash(), None).expect("Error commiting genesis state to state DB");
-			db.write(batch).expect("Error writing genesis state to state DB");
+			try!(state_db.commit(&batch, 0, &spec.genesis_header().hash(), None));
+			try!(db.write(batch).map_err(|s| ClientError::Database(s)));
```
Comment (debris) [inline ethcore/src/client/client.rs]: same here

File: ethcore/src/miner/external.rs [hunk 4]
```diff
@@ -25,67 +26,69 @@ pub trait ExternalMinerService: Send + Sync {
...
 		// given
 		let m = miner();
-		assert_eq!(m.is_mining(), false);
+		assert_eq!(m.hashrate(), U256::from(0));
+		m.submit_hashrate(U256::from(10), H256::from(1));
+		assert_eq!(m.hashrate(), U256::from(10));
 
 		// when
-		m.submit_hashrate(U256::from(10), H256::from(1));
+		sleep(Duration::from_secs(3));
```
Comment (debris) [inline ethcore/src/miner/external.rs]: This is wrong. The timer should be injected instead eg. https://github.com/debris/transient-hashmap/blob/master/src/lib.rs#L28

File: ethcore/src/client/client.rs (hunk 1 above)
Comment (gavofyork) [inline ethcore/src/client/client.rs]: `s` is a `String`, not a `DatabaseError` and implementing `ClientError::From<String>` is a little to general.

File: rpc/src/v1/impls/eth.rs (hunk 2 above)
Comment (gavofyork) [inline rpc/src/v1/impls/eth.rs]: it has different semantics - `sealing_work` being enabled means that we are preparing pending blocks, but not that any have actually been used by miners. `is_sealing` actually depends on miners calling the `eth_getWork` RPC or having it delivered to them through HTTP. usually these are more or less the same, but if e.g. we use `--force-sealing` or in the future decide to link pending-preparation with sending own transactions or querying the pending logs, then they're different.

File: ethcore/src/miner/external.rs (hunk 4 above)
Comment (gavofyork) [inline ethcore/src/miner/external.rs]: that's a lot of additional complexity in the codebase for very little gain.

**Pull request merged**